csrc/multidevice/collective.cpp

```diff
@@ -0,0 +1,199 @@
+  TORCH_INTERNAL_ASSERT(
+      std::find(
+          self.params().team.begin(),
+          self.params().team.end(),
+          comm.deviceId()) != self.params().team.end(),
+      "current device index ",
+      comm.deviceId(),
+      " must be present in the collective's team");
+}
+
+} // namespace
+
+Collective::Collective(CollectiveParams params, std::string name, bool has_root)
+    : params_(std::move(params)),
+      collective_type_(std::move(name)),
+      has_root_(has_root) {
+  assertBuffersHaveSameSize(params_.src_bufs, params_.dst_bufs);
+  TORCH_INTERNAL_ASSERT(
+      std::unique(params_.team.begin(), params_.team.end()) ==
+          params_.team.end(),
+      "the collective must not involve the same device more than once");
+  TORCH_INTERNAL_ASSERT(
+      params_.team.size() > 1, "the team size must be greater than 1");
+  if (has_root_) {
+    auto it = std::find(params_.team.begin(), params_.team.end(), params_.root);
+    TORCH_INTERNAL_ASSERT(
+        it != params_.team.end(),
+        "root (device ",
+        params_.root,
+        ") must be present in the collective's team");
+    // pytorch's process group expects the root to be specified
+    // by its relative rank, i.e., the device index's offset within the team
+    root_rank_ = std::distance(params_.team.begin(), it);
+  }
+}
+
+std::string Collective::toString(int indent) const {
+  std::stringstream ss;
+  std::string ext_indent(" ", indent);
+  std::string indent1 = ext_indent + "  ";
+  std::string indent2 = ext_indent + "    ";
+
+  ss << ext_indent << "Collective " << collective_type_ << ": {\n";
+
+  if (has_root_) {
+    ss << indent1 << "root: " << params_.root << ",\n";
+  }
+  ss << indent1 << "team: {";
+  for (auto r : params_.team) {
+    ss << r << ", ";
+  }
+  ss << indent1 << "}\n";
+  ss << indent1 << "src_bufs: {";
+  for (auto& t : params_.src_bufs) {
+    ss << "\n" << t;
+  }
+  ss << "\n" << indent1 << "}\n";
+  ss << ext_indent << "}";
+
+  return ss.str();
+}
+
+Broadcast::Broadcast(CollectiveParams params)
+    : Collective(params, "broadcast") {}
+
+c10::intrusive_ptr<c10d::Work> Broadcast::post(Communicator& comm) {
+  post_common(*this, comm);
+
+  if (comm.deviceId() == params_.root) {
+    assertBufferCount(params_.src_bufs, 1);
+    assertBufferCount(params_.dst_bufs, 0);
+  } else {
+    assertBufferCount(params_.src_bufs, 0);
+    assertBufferCount(params_.dst_bufs, 1);
+  }
+
+  return comm.getTeam(params_.team)
+      ->broadcast(
+          comm.deviceId() == params_.root ? params_.src_bufs : params_.dst_bufs,
+          {.rootRank = root_rank_});
+}
+
+Gather::Gather(CollectiveParams params) : Collective(params, "gather") {
+  assertBufferCount(params_.src_bufs, 1);
+}
+
+c10::intrusive_ptr<c10d::Work> Gather::post(Communicator& comm) {
+  post_common(*this, comm);
+
+  if (comm.deviceId() == params_.root) {
+    assertBufferCount(params_.dst_bufs, params_.team.size());
```

Comment:
So, in PyTorch, a gather generates as many buffers as the number of devices? That seems like a scalability limiter...

Reply:
Yes, this is how the pytorch process group syntax is, [check here](https://github.com/pytorch/pytorch/blob/e1aba2c8c39d817da391a75aa7ce6dd590f93755/torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp#L1502C27-L1502C27)
However the list of tensors is then flattened as one big tensor ([check here](https://github.com/pytorch/pytorch/blob/e1aba2c8c39d817da391a75aa7ce6dd590f93755/torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp#L1534)), so there is no real scalability issue

Reply:
Hmm, this doesn't seem quite efficient. Suppose we want to gather just one byte from 1K devices. We would need to get 1K 1-byte buffers, which could be a performance problem. Furthermore, unless those 1K buffers were allocated adjacently with no gap, the flattened buffer would be non-contiguous, which I suppose should be avoided for best performance.

This shouldn't matter if only a small number of devices were used, but I'm pretty sure it'd be highly inefficient for thousands of devices.

Reply:
nccl uses of "groups" to merge communications, https://github.com/pytorch/pytorch/blob/f81eacd30c4814080e5ee18fc6a53bc64ba6c5e4/torch/csrc/cuda/nccl.cpp#L1018

This should allow the communications/buffers to be fused.

Anyway, we simply follow the API needs here, the performance is up to the NCCL ProcessGroup implementation. Later we will add other backends and will run benchmarks to see the performance of each backend.

Reply:
I don't think the grouping in NCCL fuses buffers. It's just a mechanism to bundle multiple communication ops.

I understand there is nothing we can do as long as we use C10D. Can you please leave some comments somewhere that we need separate buffers for separate devices due to C10D?

Reply:
Some pytorch collective has one-buffer variant, for example:
https://pytorch.org/docs/stable/distributed.html#torch.distributed.all_gather_into_tensor
But unfortunately `gather` does not have one.

Reply:
> Can you please leave some comments somewhere that we need separate buffers for separate devices due to C10D?

I added a "NOTE" in the top comment in `multidevice/collective.h`